File: implementation/scipamato/scipamato-entity/src/test/java/ch/difty/scipamato/entity/CodeClassTest.java
package ch.difty.scipamato.entity;

import static org.assertj.core.api.Assertions.*;

import javax.validation.ConstraintViolation;

import org.junit.Test;

public class CodeClassTest extends Jsr303ValidatedEntityTest<CodeClass> {

    private static final String JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE = "{javax.validation.constraints.NotNull.message}";
    private static final String THIS_IS_CC1 = "this is cc1";

    private void verifySuccessfulValidation(final CodeClass cc) {
        validate(cc);
        assertThat(getViolations()).isEmpty();
    }

    private void validateAndAssertFailure(final CodeClass cc, final String field, final Object invalidValue, final String msg) {
        validate(cc);

        assertThat(getViolations()).isNotEmpty().hasSize(1);
        ConstraintViolation<CodeClass> violation = getViolations().iterator().next();
        assertThat(violation.getMessageTemplate()).isEqualTo(msg);
        assertThat(violation.getInvalidValue()).isEqualTo(invalidValue);
        assertThat(violation.getPropertyPath().toString()).isEqualTo(field);
    }

    @Test
    public void validatingCodeClass_beingValid_succeeds() {
        verifySuccessfulValidation(new CodeClass(1, "foo", "bar"));
    }

    @Test
    public void validatingCodeClass_withNullName_fails() {
        CodeClass cc = new CodeClass(1, null, "bar");
        validateAndAssertFailure(cc, CodeClass.NAME, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE);
    }

    @Test
    public void validatingCodeClass_withNullDescription_fails() {
        CodeClass cc = new CodeClass(1, "foo", null);
        validateAndAssertFailure(cc, CodeClass.DESCRIPTION, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE);
    }

    @Test
    public void testingToString() {
        CodeClass cc = new CodeClass(1, "foo", "bar");
        assertThat(cc.toString()).isEqualTo("CodeClass[id=1]");
    }

    @Test
    public void cloning_copiesValues() {
        CodeClass orig = new CodeClass(1, "cc1", THIS_IS_CC1);
        CodeClass copy = new CodeClass(orig);
        assertThat(copy).isEqualToComparingFieldByField(orig);
    }

    @Test
    public void sameValues_makeEquality() {
        CodeClass cc1 = new CodeClass(1, "cc1", THIS_IS_CC1);
        CodeClass cc2 = new CodeClass(cc1);
        assertEquality(cc1, cc2);
    }

    private void assertEquality(CodeClass cc1, CodeClass cc2) {
        assertThat(cc1.equals(cc2)).isTrue();
        assertThat(cc2.equals(cc1)).isTrue();
        assertThat(cc1.hashCode()).isEqualTo(cc2.hashCode());
    }

    @Test
    public void differingValues_makeInequality() {
        CodeClass cc1 = new CodeClass(1, "cc1", THIS_IS_CC1);
        CodeClass cc2 = new CodeClass(2, "cc1", THIS_IS_CC1);
        CodeClass cc3 = new CodeClass(1, "cc2", THIS_IS_CC1);
        CodeClass cc4 = new CodeClass(1, "cc1", "this is cc2");

        assertThat(cc1.equals(cc2)).isFalse();
        assertThat(cc1.equals(cc3)).isFalse();
        assertThat(cc1.equals(cc4)).isFalse();
        assertThat(cc2.equals(cc3)).isFalse();
        assertThat(cc2.equals(cc4)).isFalse();
        assertThat(cc3.equals(cc4)).isFalse();

        assertThat(cc1.hashCode()).isNotEqualTo(cc2.hashCode());
        assertThat(cc1.hashCode()).isNotEqualTo(cc3.hashCode());
        assertThat(cc1.hashCode()).isNotEqualTo(cc4.hashCode());
        assertThat(cc2.hashCode()).isNotEqualTo(cc3.hashCode());
        assertThat(cc2.hashCode()).isNotEqualTo(cc4.hashCode());
        assertThat(cc3.hashCode()).isNotEqualTo(cc4.hashCode());
    }

    @SuppressWarnings("unlikely-arg-type")
    @Test
    public void equalingToSpecialCases() {
        CodeClass cc1 = new CodeClass(1, "cc1", THIS_IS_CC1);

        assertThat(cc1.equals(cc1)).isTrue();
        assertThat(cc1.equals(null)).isFalse();
        assertThat(cc1.equals("")).isFalse();
    }

    @Test
    public void displayValue() {
        CodeClass cc1 = new CodeClass(1, "cc1", THIS_IS_CC1);
        assertThat(cc1.getDisplayValue()).isEqualTo("cc1");
    }

    @Test
    public void differingValues_withIdNullOnOne() {
        CodeClass cc1 = new CodeClass(1, "cc1", THIS_IS_CC1);
        CodeClass cc2 = new CodeClass(null, "cc1", THIS_IS_CC1);
        assertInequality(cc1, cc2);
    }

    private void assertInequality(CodeClass cc1, CodeClass cc2) {
        assertThat(cc1.equals(cc2)).isFalse();
        assertThat(cc2.equals(cc1)).isFalse();
        assertThat(cc1.hashCode()).isNotEqualTo(cc2.hashCode());
    }

    @Test
    public void differingValues_withIdNullOnBoth() {
        CodeClass cc1 = new CodeClass(null, "cc1", THIS_IS_CC1);
        CodeClass cc2 = new CodeClass(null, "cc1", THIS_IS_CC1);
        assertEquality(cc1, cc2);
    }

    @Test
    public void differingValues_withNameNullOnOne() {
        CodeClass cc1 = new CodeClass(1, null, THIS_IS_CC1);
        CodeClass cc2 = new CodeClass(1, "cc1", THIS_IS_CC1);
        assertInequality(cc1, cc2);
    }

    @Test
    public void differingValues_withNameNullOnBoth() {
        CodeClass cc1 = new CodeClass(1, null, THIS_IS_CC1);
        CodeClass cc2 = new CodeClass(1, null, THIS_IS_CC1);
        assertEquality(cc1, cc2);
    }

    @Test
    public void differingValues_withDescriptionNullOnOne() {
        CodeClass cc1 = new CodeClass(1, "cc1", THIS_IS_CC1);
        CodeClass cc2 = new CodeClass(1, "cc1", null);
        assertInequality(cc1, cc2);
    }

    @Test
    public void differingValues_withDescriptionNullOnBoth() {
        CodeClass cc1 = new CodeClass(1, "cc1", null);
        CodeClass cc2 = new CodeClass(1, "cc1", null);
        assertEquality(cc1, cc2);
    }
}

File: core/core-entity/src/test/kotlin/ch/difty/scipamato/core/entity/CodeClassTest.kt
package ch.difty.scipamato.core.entity

import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldNotBeEqualTo
import org.junit.jupiter.api.Test

internal class CodeClassTest : Jsr303ValidatedEntityTest<CodeClass>(CodeClass::class.java) {

    override fun newValidEntity(): CodeClass {
        return CodeClass(1, "foo", "bar")
    }

    override val toString = "CodeClass[id=1]"
    override val displayValue = "${newValidEntity().id} - ${newValidEntity().name}"

    @Test
    fun validatingCodeClass_beingValid_succeeds() {
        verifySuccessfulValidation(CodeClass(1, "foo", "bar"))
    }

    @Test
    fun cloning_copiesValues() {
        val orig = CodeClass(1, "cc1", DESC)
        val copy = CodeClass(orig)
        copy shouldBeEqualTo orig
    }

    @Test
    fun sameValues_makeEquality() {
        val cc1 = CodeClass(1, "cc1", DESC)
        val cc2 = CodeClass(cc1)
        assertEquality(cc1, cc2)
    }

    private fun assertEquality(cc1: CodeClass, cc2: CodeClass) {
        (cc1 == cc2).shouldBeTrue()
        (cc2 == cc1).shouldBeTrue()
        cc1.hashCode() shouldBeEqualTo cc2.hashCode()
    }

    @Test
    fun differingValues_makeInequality() {
        val cc1 = CodeClass(1, "cc1", DESC)
        val cc2 = CodeClass(2, "cc1", DESC)
        val cc3 = CodeClass(1, "cc2", DESC)
        val cc4 = CodeClass(1, "cc1", "this is cc2")

        (cc1 == cc2).shouldBeFalse()
        (cc1 == cc3).shouldBeFalse()
        (cc1 == cc4).shouldBeFalse()
        (cc2 == cc3).shouldBeFalse()
        (cc2 == cc4).shouldBeFalse()
        (cc3 == cc4).shouldBeFalse()

        cc1.hashCode() shouldNotBeEqualTo cc2.hashCode()
        cc1.hashCode() shouldNotBeEqualTo cc3.hashCode()
        cc1.hashCode() shouldNotBeEqualTo cc4.hashCode()
        cc2.hashCode() shouldNotBeEqualTo cc3.hashCode()
        cc2.hashCode() shouldNotBeEqualTo cc4.hashCode()
        cc3.hashCode() shouldNotBeEqualTo cc4.hashCode()
    }

    @Test
    fun equalingToSpecialCases() {
        val cc1 = CodeClass(1, "cc1", DESC)

        (cc1 == cc1).shouldBeTrue()
    }

    @Test
    fun differingValues_withIdNullOnOne() {
        val cc1 = CodeClass(1, "cc1", DESC)
        val cc2 = CodeClass(null, "cc1", DESC)
        assertInequality(cc1, cc2)
    }

    private fun assertInequality(cc1: CodeClass, cc2: CodeClass) {
        (cc1 == cc2).shouldBeFalse()
        (cc2 == cc1).shouldBeFalse()
        cc1.hashCode() shouldNotBeEqualTo cc2.hashCode()
    }

    @Test
    fun differingValues_withIdNullOnBoth() {
        val cc1 = CodeClass(null, "cc1", DESC)
        val cc2 = CodeClass(null, "cc1", DESC)
        assertEquality(cc1, cc2)
    }

    companion object {
        private const val DESC = "this is cc1"
    }
}
